### ISSUE TITLE
mammouth<->PHP

### ISSUE THREAD

**Issue by aolko** [btwael/mammouth#58]
see http://js2.coffee/

Basically Mammouth -> PHP & PHP-> Mammouth online conversion

**Comment by coderofsalvation** [comment]
exactly my thought.
Without this projects won't be converted to mammouth.

**Comment by btwael** [comment]
But why you want really to convert PHP to mammouth?

**Comment by Manaus** [comment]
Maintainability of already written code, I would say (I do it often with js<->coffee)

**Comment by btwael** [comment]
The problematic of converting PHP to Mammouth, it's that Mammouth is more expressive than PHP, by example loops in Mammouth are used to iterate over arrays, while php ones are used to repeat a peace of instuctions.
So even if we implement this PHP to Mammouth, the output mammouth code will be very long compared to a hand-written Mammouth code.

**Comment by coderofsalvation** [comment]
btwael: Long php code is not a problem. Syntactical lock-in (now) is a problem.

In some eyes coffeescript was a far more superior and readable than javascript.
However for the haters, coffeescript was just useless syntactical sugar.
So why are there coffeescript files found in so many npm modules?
The answer is: Syntactic interoperability, developers can choose between writing .js or .coffee.
For every .coffee file, there's an .js file.
JS devs write .js files, and coffeescript devs write .coffee files (and run `coffee -c *.coffee` before pushing their repo)
So when the maintainer receives a pullrequest with a .js file, he can convert it to coffeescript (and vice versa).

As long as mammouth its an one-way converter, it causes syntactical lock-in.
The 'sell it to your boss'-chance is very low.
Create a two-way converter and nobody can argue against mamouth because *mammouth would be PHP, and PHP can be mammouth*

Mammouth's portability could even become bigger if you would add a --coffeescript compatibility flag to mammouth.

**Comment by CezaryDanielNowak** [comment]
@coderofsalvation 
- It's unlikely to convert js code to coffeescript without breaking anything. Even js generated by coffeescript has some trouble.
- pushing transpiled code (written in any language) is bad practice, it is not common practice for sure.

I personally used js->cofee just to limit my work with removing curly brackets and semicolons.

**Comment by coderofsalvation** [comment]
Hi CezaryDanielNowak. Thanx for stepping in. 

Im having trouble to accept your statements without having a why or what mentioned.
Could you be so kind to elaborate on that? Thx

> It's unlikely to convert js code to coffeescript without breaking anything. Even js generated by coffeescript has some trouble.

I've never heard/experienced trouble with js generated by coffeescript other than reference issues in angularjs (but that's got a workaround, and happens due to angular's [preprocessor](https://github.com/Plateful/plateful-mobile/wiki/AngularJS-CoffeeScript-Style-Guide#controllers ).
Maybe you're talking about the early days of coffeescript?

> pushing transpiled code (written in any language) is bad practice, it is not common practice for sure.
You're not telling why, so I have to disagree with your personal taste.

**Comment by soulcreate** [comment]
+1

**Comment by coderofsalvation** [comment]
*bump*

**Comment by aolko** [comment]
dead?

**Comment by btwael** [comment]
Hey everybody! Your comments and the download statistics in npm, made me rethink about the future of Mammouth, and I just released (after experience developing compilers in C/C++) that I can create something interesting whatever the future of PHP. So I decided to start coding a new version of mammouth that will support different version of PHP (so you can generate code compatible with whatever PHP version) and import and use already written PHP code with efficiency.

How that can affect this issue? An analyser (parser + ...) for PHP will be developed so we can import PHP, once this is done, I can easily create compile PHP to Mammouth.

You can contribute to Mammouth new version in this [branch](https://github.com/btwael/mammouth/tree/v4.x).

Yu can also support me financially doing so on [Patreon](https://www.patreon.com/btwael).

**Comment by aolko** [comment]
Are you aware that there's PHP parser in php nowadays? (just checking)

**Comment by btwael** [comment]
Yes, I am aware PHP parser (written in PHP), but Mammouth is developed in Javascript, so it's better to develop our parser in javascript.

**Comment by btwael** [comment]
I thought you were talking about [this one](https://github.com/nikic/PHP-Parser), I will check this one so if it's well compatible with our design concept, we will depend on it, otherwise we will create our parser (the goal is to create a parser that support different PHP versions).

**Comment by aolko** [comment]
can you also do a gem? (for the people who can't into grunt/gulp/other burnt-out build system)

**Comment by btwael** [comment]
Depend on how advanced the job we did, we may create a CLI tool that automates the compilation job.

**Comment by aolko** [comment]
ideally a tool that can be used as file watcher ([what's that?](https://www.jetbrains.com/help/phpstorm/file-watchers.html)) would be nice
